cmd/snap-confine/mount-support-hybris.c

```diff
@@ -0,0 +1,198 @@
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <stdint.h>
+#include <unistd.h>
+/* POSIX version of basename() and dirname() */
+#include <libgen.h>
+
+#include "../libsnap-confine-private/classic.h"
+#include "../libsnap-confine-private/cleanup-funcs.h"
+#include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/utils.h"
+#include "mount-support.h"
+
+#define SC_HYBRIS_PROPERTY_FILE "/system/build.prop"
+
+#define SC_LIBGL_DIR   SC_EXTRA_LIB_DIR "/gl"
+#define SC_VULKAN_DIR  SC_EXTRA_LIB_DIR "/vulkan"
+#define SC_GLVND_DIR  SC_EXTRA_LIB_DIR "/glvnd"
+
+#define SC_VULKAN_SOURCE_DIR "/usr/share/vulkan"
+#define SC_EGL_VENDOR_SOURCE_DIR "/usr/share/glvnd"
+
+#define SC_HYBRIS_ROOTFS "/android"
+#define SC_HYBRIS_SYSTEM_SYMLINK "/system"
+#define SC_HYBRIS_VENDOR_SYMLINK "/vendor"
+#define SC_HYBRIS_ODM_SYMLINK "/odm"
+#define SC_HYBRIS_APEX_SYMLINK "/apex"
+#define SC_HYBRIS_SYSTEM_SYMLINK_TARGET "/android/system"
+#define SC_HYBRIS_VENDOR_SYMLINK_TARGET "/android/vendor"
+#define SC_HYBRIS_ODM_SYMLINK_TARGET "/android/odm"
+#define SC_HYBRIS_APEX_SYMLINK_TARGET "/android/apex"
+
+static const char *hybris_globs[] = {
+	"libEGL_libhybris.so*",
+	"libGLESv1_CM_libhybris.so*",
+	"libGLESv2_libhybris.so*",
+	"libhybris-common.so*",
+	"libhybris-platformcommon.so*",
+	"libhybris-eglplatformcommon.so*",
+	"libgralloc.so*",
+	"libsync.so*",
+	"libhardware.so*",
+	"libui.so*",
+	"libhybris/eglplatform_*.so",
+	"libhybris/linker/*.so"
+};
+
+static const size_t hybris_globs_len =
+    sizeof hybris_globs / sizeof *hybris_globs;
+
+// Location for libhybris vulkan files (including _wayland)
+static const char *hybris_vulkan_globs[] = {
+	"icd.d/*hybris*.json",
+};
+
+static const size_t hybris_vulkan_globs_len =
+    sizeof hybris_vulkan_globs / sizeof *hybris_vulkan_globs;
+
+// Location of EGL vendor files
+static const char *hybris_egl_vendor_globs[] = {
+	"egl_vendor.d/*hybris*.json",
+};
+
+static const size_t hybris_egl_vendor_globs_len =
+    sizeof hybris_egl_vendor_globs / sizeof *hybris_egl_vendor_globs;
+
+static void sc_hybris_mount_android_rootfs(const char *rootfs_dir)
+{
+	// Bind mount a tmpfs on $rootfs_dir/$tgt_dir (i.e. /var/lib/snapd/lib/gl)
+	char path_buf[PATH_MAX] = { 0 };
+	sc_must_snprintf(path_buf, sizeof(path_buf), "%s%s", rootfs_dir, SC_HYBRIS_ROOTFS);
+	const char *android_rootfs_dir = path_buf;
+
+	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+	int res = mkdir(android_rootfs_dir, 0755);
+	if (res != 0 && errno != EEXIST) {
+		die("cannot create tmpfs target %s", android_rootfs_dir);
+	}
+	if (res == 0 && (chown(android_rootfs_dir, 0, 0) < 0)) {
+		// Adjust the ownership only if we created the directory.
+		die("cannot change ownership of %s", android_rootfs_dir);
+	}
+	(void)sc_set_effective_identity(old);
+
+	(void)mount(SC_HYBRIS_ROOTFS, android_rootfs_dir, NULL, MS_BIND | MS_REC | MS_RDONLY, NULL);
```

Comment:
Should this be checked and should s-c die if it does not work? If it is legitimately allowed to fail, as indicated by the `(void)` up front, can you add a comment just above
to explain why.

Reply:
Can take a deeper look later after `$DAYJOB`

Reply:
Handled in d5a85d6